presage-store-sqlite/src/protocol.rs

```diff
@@ -262,18 +595,48 @@ impl SenderKeyStore for SqliteProtocolStore {
         &mut self,
         sender: &ProtocolAddress,
         distribution_id: Uuid,
-        // TODO: pass this by value!
         record: &SenderKeyRecord,
-    ) -> Result<(), ProtocolError> {
-        todo!()
+    ) -> Result<(), SignalProtocolError> {
+        let address = sender.name();
+        let device_id: u32 = sender.device_id().into();
+        let record = record.serialize()?;
+        query!(
+            "INSERT INTO sender_keys
+            (address, device_id, identity, distribution_id, record)
+            VALUES (?1, ?2, ?3, ?4, ?5)
+            ON CONFLICT DO UPDATE SET record = ?5",
+            address,
+            device_id,
+            self.identity,
+            distribution_id,
+            record,
+        )
+        .execute(&self.store.db)
+        .await
+        .into_protocol_error()?;
+        Ok(())
     }
 
     /// Look up the entry corresponding to `(sender, distribution_id)`.
     async fn load_sender_key(
         &mut self,
         sender: &ProtocolAddress,
         distribution_id: Uuid,
-    ) -> Result<Option<SenderKeyRecord>, ProtocolError> {
-        todo!()
+    ) -> Result<Option<SenderKeyRecord>, SignalProtocolError> {
+        let address = sender.name();
+        let device_id: u32 = sender.device_id().into();
+        query_scalar!(
+            "SELECT record FROM sender_keys
+            WHERE address = ? AND device_id = ? AND identity = ? AND distribution_id = ?",
+            address,
+            device_id,
+            self.identity,
+            distribution_id,
+        )
+        .fetch_optional(&self.store.db)
+        .await
+        .into_protocol_error()?
+        .map(|record| SenderKeyRecord::deserialize(&record))
+        .transpose()
```

Comment:
I think we can make this actually more performant, but deserializing directly from the sqlite buffer.

Reply:
Do you mean that there's a way with `sqlx` to get a `&[u8]` instead for such BLOB fields?